pkg/controller/ksmtuned/controller.go

```diff
@@ -0,0 +1,148 @@
+package ksmtuned
+
+import (
+	"context"
+	"reflect"
+
+	"github.com/sirupsen/logrus"
+
+	ksmtunedv1 "github.com/harvester/node-manager/pkg/apis/node.harvesterhci.io/v1beta1"
+	ctlnode "github.com/harvester/node-manager/pkg/generated/controllers/core/v1"
+	ctlksmtuned "github.com/harvester/node-manager/pkg/generated/controllers/node.harvesterhci.io/v1beta1"
+	"github.com/harvester/node-manager/pkg/ksmtuned"
+	"github.com/harvester/node-manager/pkg/metrics"
+)
+
+const (
+	HandlerName     = "harvester-ksmtuned-handler"
+	NodeHandlerName = "harvester-ksmtuned-node-handler"
+)
+
+var (
+	modes = map[ksmtunedv1.KsmtunedMode]ksmtunedv1.KsmtunedParameters{
+		ksmtunedv1.StandardMode: {
+			SleepMsec: 20,
+			Boost:     0,
+			Decay:     0,
+			MinPages:  100,
+			MaxPages:  100,
+		},
+		ksmtunedv1.HighMode: {
+			SleepMsec: 20,
+			Boost:     200,
+			Decay:     50,
+			MinPages:  100,
+			MaxPages:  10000,
+		},
+	}
+)
+
+type Controller struct {
+	ctx      context.Context
+	NodeName string
+
+	KsmtunedCache ctlksmtuned.KsmtunedCache
+	Ksmtuneds     ctlksmtuned.KsmtunedController
+
+	NodeCache ctlnode.NodeCache
+	Nodes     ctlnode.NodeController
+
+	Ksmtuned *ksmtuned.Ksmtuned
+
+	CurrentKsmtuendSpec *ksmtunedv1.KsmtunedSpec
+}
+
+func Register(ctx context.Context, nodeName string, kts ctlksmtuned.KsmtunedController, nodes ctlnode.NodeController) (*Controller, error) {
+	k, err := ksmtuned.NewKsmtuned(ctx, nodeName)
+	if err != nil {
+		return nil, err
+	}
+
+	k.SetKsmdUtilization(metrics.KsmdUtilizationGV)
+	c := &Controller{
+		ctx:           ctx,
+		NodeName:      nodeName,
+		KsmtunedCache: kts.Cache(),
+		Ksmtuneds:     kts,
+		NodeCache:     nodes.Cache(),
+		Nodes:         nodes,
+		Ksmtuned:      k,
+	}
+
+	c.Ksmtuneds.OnChange(ctx, HandlerName, c.OnChange)
+	c.Ksmtuneds.OnRemove(ctx, HandlerName, c.OnRemove)
+
+	c.Nodes.OnChange(ctx, NodeHandlerName, c.NodeOnChange)
+
+	go c.watchStatus(nodeName)
+
+	return c, nil
+}
+
+func (c *Controller) OnChange(key string, kt *ksmtunedv1.Ksmtuned) (*ksmtunedv1.Ksmtuned, error) {
+	if kt == nil || kt.DeletionTimestamp != nil {
+		return nil, nil
+	} else if key != c.NodeName {
+		return kt, nil
+	}
+
+	var (
+		parameters ksmtunedv1.KsmtunedParameters
+		ok         bool
+	)
+
+	if kt.Spec.Run == ksmtunedv1.Stop {
+		return kt, c.Ksmtuned.Stop()
+	} else if kt.Spec.Run == ksmtunedv1.Prune {
+		return kt, c.Ksmtuned.Prune()
+	} else if parameters, ok = modes[kt.Spec.Mode]; !ok { // customized mode
+		return kt, c.Ksmtuned.Apply(kt.Spec.ThresCoef, kt.Spec.KsmtunedParameters)
+	}
+
+	if !reflect.DeepEqual(kt.Spec.KsmtunedParameters, parameters) {
+		newObj := kt.DeepCopy()
+		newObj.Spec.KsmtunedParameters = parameters
+
+		if _, err := c.Ksmtuneds.Update(newObj); err != nil {
+			return nil, err
+		}
+		kt = newObj
+	}
+	return kt, c.Ksmtuned.Apply(kt.Spec.ThresCoef, parameters)
```

Comment:
shoud we apply first and then update the ksm object since it will reconcile if contains any errors and the object will not be updated incorrect if the apply fails later.